File: my_flutter_gen/lib/feature/auth/data/repository/auth_repository_impl.dart
import 'package:my_flutter_gen/core/error/error_handler.dart';
import 'package:my_flutter_gen/core/network/api_client.dart';
import 'package:my_flutter_gen/core/network/api_result.dart';
import 'package:my_flutter_gen/feature/auth/data/local/local_data_source.dart';
import 'package:my_flutter_gen/feature/auth/data/remote/model/login_response.dart';
import 'package:my_flutter_gen/feature/auth/data/remote/remote_data_source.dart';
import 'package:my_flutter_gen/feature/auth/domain/model/user.dart';
import 'package:my_flutter_gen/feature/auth/domain/repository/auth_repository.dart';

/// AuthRepository 的具体实现
class AuthRepositoryImpl implements AuthRepository {
  final RemoteDataSource _remoteDataSource;
  final LocalDataSource _localDataSource;

  const AuthRepositoryImpl({
    required RemoteDataSource remoteDataSource,
    required LocalDataSource localDataSource,
  })  : _remoteDataSource = remoteDataSource,
        _localDataSource = localDataSource;

  @override
  Future<ApiResult<User>> login({
    required String email,
    required String password,
  }) {
    return ErrorHandler.guardAsync(() async {
      final response = await _remoteDataSource.login(
        email: email,
        password: password,
      );
      final user = _mapResponseToUser(response);

      // 缓存用户 & 设置 Token
      await _localDataSource.cacheUser(user);
      ApiClient.instance.setToken(response.token);

      return user;
    });
  }

  @override
  Future<ApiResult<User>> register({
    required String username,
    required String email,
    required String password,
  }) {
    return ErrorHandler.guardAsync(() async {
      final response = await _remoteDataSource.register(
        username: username,
        email: email,
        password: password,
      );
      final user = _mapResponseToUser(response);

      await _localDataSource.cacheUser(user);
      ApiClient.instance.setToken(response.token);

      return user;
    });
  }

  @override
  Future<ApiResult<void>> logout() {
    return ErrorHandler.guardAsync(() async {
      await _remoteDataSource.logout();
      await _localDataSource.clearCache();
      ApiClient.instance.clearToken();
    });
  }

  @override
  Future<User?> getCachedUser() {
    return _localDataSource.getCachedUser();
  }

  @override
  Future<bool> isLoggedIn() async {
    final user = await _localDataSource.getCachedUser();
    return user?.token != null;
  }

  /// DTO → Domain Model 映射
  User _mapResponseToUser(LoginResponse response) {
    return User(
      id: response.id,
      username: response.username,
      email: response.email,
      avatarUrl: response.avatarUrl,
      token: response.token,
    );
  }
}

File: MyComposeProject/app/src/main/java/com/louis/mycomposeproject/data/repository/AuthRepositoryImpl.kt
package com.louis.mycomposeproject.data.repository

import com.louis.mycomposeproject.data.local.PreferencesManager
import com.louis.mycomposeproject.data.remote.AuthApi
import com.louis.mycomposeproject.data.remote.LoginRequest
import com.louis.mycomposeproject.data.remote.RegisterRequest
import com.louis.mycomposeproject.domain.model.User
import com.louis.mycomposeproject.domain.repository.AuthRepository
import com.louis.mycomposeproject.util.NetworkErrorHandler

class AuthRepositoryImpl(
    private val api: AuthApi,
    private val preferencesManager: PreferencesManager
) : AuthRepository {
    override suspend fun login(email: String, password: String): Result<User> {
        return try {
            val response = api.login(LoginRequest(email, password))
            val user = User(response.id, response.email, response.token)
            preferencesManager.saveToken(user.token)
            preferencesManager.saveUserId(user.id)
            preferencesManager.saveEmail(user.email)
            Result.success(user)
        } catch (e: Exception) {
            Result.failure(Exception(NetworkErrorHandler.getErrorMessage(e)))
        }
    }
    
    override suspend fun register(email: String, password: String, name: String): Result<User> {
        return try {
            val response = api.register(RegisterRequest(email, password, name))
            val user = User(response.id, response.email, response.token)
            Result.success(user)
        } catch (e: Exception) {
            Result.failure(Exception(NetworkErrorHandler.getErrorMessage(e)))
        }
    }
    
    override fun logout() {
        preferencesManager.clearAll()
    }
    
    override fun isLoggedIn(): Boolean {
        return preferencesManager.isLoggedIn()
    }
}
